add workflow's middlewares

ampelmann opened the pull request:
Refs #73 

Please review @jbbarth @ybastide @benjastudio

ybastide commented on simpleflow/executor.py:
```diff
@@ -118,3 +111,24 @@ def fail(self, reason, details=None):
+        return self._workflow.after_finished(self._history)
+
+    def after_run(self):
+        """
```
`warnings.warn("deprecated", PendingDeprecationWarning)`? (brownie points: as a decorator, changing the filter level, etc).

ybastide commented on simpleflow/executor.py:
```diff
@@ -118,3 +111,24 @@ def fail(self, reason, details=None):
+        """
+        return self.after_finished()
+
+    def before_run(self):
```
Already defined (just above `run`)

jbbarth commented on simpleflow/swf/executor.py:
```diff
@@ -367,8 +369,10 @@ def replay(self, history):
 
             return [decision], {}
 
+        self.after_replay()
```
Cosmetic: why separate the two calls?

jbbarth replied:
Nevermind, forget that.

jbbarth commented:
Looks good to me, same as @ybastide I'd better have real warnings for unused deprecated methods.

ampelmann commented:
@jbbarth @ybastide Job done !

ampelmann commented:
@ybastide done

ybastide commented:
Thanks!

ampelmann commented:
Some changes : 
* I renamed after_finished by after_closed (the SWF term)
* I added `on_completed` that is called only when the workflow succeeded.

Ping @jbbarth

jbbarth commented:
The branch needs to be rebased agaist master, won't merge as is.

jbbarth commented on simpleflow/executor.py:
```diff
@@ -116,8 +122,8 @@ def before_replay(self):
-    def after_finished(self):
-        return self._workflow.after_finished(self._history)
+    def after_closed(self):
+        return self._workflow.after_closed(self._history)
```
There are other occurrences of `after_finished()` left, they should be renamed too.
